trans/train.py

```diff
@@ -322,15 +323,31 @@ def train_worker_init_fn(worker_id):
     best_epoch = 0
     patience = 0
 
+    sample_ids = list(range(len(training_data.samples)))
+    sample_stack = []
+
     for epoch in range(args.epochs):
 
         logging.info("Training...")
         transducer_.train()
         transducer_.zero_grad()
         with utils.Timer():
             train_loss = 0.
-            # rollin not implemented at the moment
-            # rollin = rollin_schedule(epoch)
+
+            #rollin
+            if args.rollin is not None and args.rollin:
+                rollin = rollin_schedule(epoch)
+                nr_samples = int(len(training_data.samples) * rollin)
+                rollin_samples = random.sample(sample_ids, nr_samples)
+                with torch.no_grad():
+                    # restore
+                    for id_, sample in sample_stack:
+                        training_data.samples[id_] = sample
+                    sample_stack = []
+                    for id_ in rollin_samples:
+                        sample_stack.append((id_, dataclasses.replace(training_data.samples[id_])))
+                        transducer_.roll_in(training_data.samples[id_], rollin)
+
             j = 0
             for j, batch in enumerate(training_data_loader):
```

Comment:
With the current setup, rollin is performed before each epoch. I think a problem with this approach could be that the model performs update during the epoch and the rolled in target sequences are not representative anymore for the model. So maybe it would make more sense to rollin after every training step with some probability.